deployment/ec2-with-asg.tf

```diff
@@ -0,0 +1,162 @@
+    DATABASE_URL          = data.dotenv.env_file.env["DATABASE_URL"]
+    SECRET_KEY_BASE       = data.dotenv.env_file.env["SECRET_KEY_BASE"]
+    POOL_SIZE             = data.dotenv.env_file.env["POOL_SIZE"]
+    BEARER_TOKEN          = data.dotenv.env_file.env["BEARER_TOKEN"]
+    PORT                  = data.dotenv.env_file.env["PORT"]
+  }))
+
+  network_interfaces {
+    associate_public_ip_address = false
+    subnet_id                   = aws_subnet.subnet_2.id
+    security_groups             = [aws_security_group.sg_for_ec2.id]
+  }
+
+  tag_specifications {
+    resource_type = "instance"
+    tags = {
+      Name = "${local.environment_prefix}ec2"
+    }
+  }
+
+  key_name = "AvantiFellows"
+
+  iam_instance_profile {
+    name = aws_iam_instance_profile.ec2_profile.name
+  }
+
+  metadata_options {
+    http_endpoint               = "enabled"
+    http_put_response_hop_limit = 1
+    http_tokens                 = "optional"
+  }
+}
+
+resource "aws_autoscaling_group" "asg" {
+  name_prefix      = "${local.environment_prefix}asg"
+  desired_capacity = 1
+  max_size         = 1
+  min_size         = 1
+
+  # connect to the target group
+  target_group_arns = [aws_lb_target_group.alb_tg.arn]
+
+  vpc_zone_identifier = [aws_subnet.subnet_2.id]
+
+  launch_template {
+    id      = aws_launch_template.ec2_launch_templ.id
+    version = "$Latest"
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "high_cpu_alarm" {
+  alarm_name          = "${local.environment_prefix}-high-cpu-alarm"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "2"
+  metric_name         = "CPUUtilization"
+  namespace           = "AWS/EC2"
+  period              = "60" # Seconds (2 minutes)
+  statistic           = "Average"
+  threshold           = "30" # Percentage
+
+  dimensions = {
+    AutoScalingGroupName = aws_autoscaling_group.asg.name
+  }
+
+  alarm_description = "This metric monitors high CPU utilization on EC2 instances"
+  alarm_actions     = [aws_autoscaling_policy.high_cpu_policy.arn]
+}
+
+resource "aws_autoscaling_policy" "high_cpu_policy" {
+  name                   = "${local.environment_prefix}-high-cpu-policy"
+  scaling_adjustment     = 1 # Increase the desired capacity by 1
+  adjustment_type        = "ChangeInCapacity"
+  cooldown               = 300 # Seconds (5 minutes)
+  autoscaling_group_name = aws_autoscaling_group.asg.name
+}
+
+# Bastion Host Instance
+resource "aws_instance" "bastion_host" {
+  ami             = "ami-05e00961530ae1b55"
+  instance_type   = "t2.micro"
+  key_name        = "AvantiFellows"
+  subnet_id       = aws_subnet.subnet_1.id # Place in a public subnet
+  security_groups = [aws_security_group.sg_bastion.id]
+
+  tags = {
+    Name = "${local.environment_prefix}Bastion-Host"
+  }
+
+  iam_instance_profile = aws_iam_instance_profile.ec2_profile.name
+
+  provisioner "file" {
+    source      = "C:/Users/amanb/.ssh/AvantiFellows.pem"
```

Comment:
this should be in a variable, and should be used like `data.dotenv` because it could be different for other people.

Reply:
Done